subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java

```diff
@@ -73,32 +67,30 @@
         }
     };
 
-    private volatile boolean buildRunning;
-
-    public WatchingVirtualFileSystem(
+    public DefaultFileSystemWatchingHandler(
         FileWatcherRegistryFactory watcherRegistryFactory,
-        AbstractVirtualFileSystem delegate,
+        AtomicSnapshotHierarchyReference snapshotHierarchyReference,
         DelegatingDiffCapturingUpdateFunctionDecorator delegatingUpdateFunctionDecorator,
         Predicate<String> watchFilter,
-        DaemonDocumentationIndex daemonDocumentationIndex
+        DaemonDocumentationIndex daemonDocumentationIndex,
+        LocationsUpdatedByCurrentBuild locationsUpdatedByCurrentBuild
     ) {
-        super(delegate);
         this.watcherRegistryFactory = watcherRegistryFactory;
+        this.snapshotHierarchyReference = snapshotHierarchyReference;
         this.delegatingUpdateFunctionDecorator = delegatingUpdateFunctionDecorator;
         this.watchFilter = watchFilter;
         this.daemonDocumentationIndex = daemonDocumentationIndex;
+        this.locationsUpdatedByCurrentBuild = locationsUpdatedByCurrentBuild;
     }
 
     @Override
     public void afterBuildStarted(boolean watchingEnabled) {
         reasonForNotWatchingFiles = null;
-        getRoot().update(currentRoot -> {
+        snapshotHierarchyReference.update(currentRoot -> {
```

Comment:
I think this name is confusing. What we are updating is the `root`, or the reference to the root. I think this would read a lot nicer:

```suggestion
        root.update(currentRoot -> {
```

Reply:
I renamed `snapshotHierarchyReference` to `root`.